runtime/cudaq/algorithms/state.cpp

```diff
@@ -39,7 +39,12 @@ std::complex<double> state::operator[](std::size_t idx) {
   if (shape.size() != 1)
     throw std::runtime_error("Cannot request 1-d index into density matrix. "
                              "Must be a state vector.");
-  return stateData[idx];
+  std::size_t numQubits = std::log2(stateData.size());
```

Comment:
Would you mind providing an you explain the updates here? Why the need to transform the index?

Reply:
**Short explanation**: Otherwise this would be a breaking change from the users perspective. This PR changes the qubit ordering for the internal representation of the state. There is a need for further clarification about the `cudaq::state` API.

**Long explanation**: There is an ambiguity on what `operator[](std::size_t i)` of `cudaq::state` should return. To illustrate the problem, suppose we define a two qubit state vector:
```
cudaq::state s = {a, b, c, d}
```
The question here is: What is the value of `s[1]`?

Well, it depends of what `1` represents. If `1` is a index to the underlying representation of the state, say a `std::vector` or `std::array`, then we would return `s[1] == b`. This interpretation, however, leaks internal details about the representation of the state. What if we eventually decide to some other underlying representation, e.g., allow `cudaq::state` to also represent sparse states?

Okay, so interpreting `1` as the index to the underlying representation is not a good idea. So we decide to say that `1` is the quantum state `|1>`. Now qubit ordering comes in to play: the state `|1>` can be represented as either `|10>` or `|01>`, depending on the chosen order. Currently, users seem to expect `|1> == |10>`, meaning that the binary string that represents `|1>` does not correspond to `1` in binary (and thus it also does correspond to the index `1` on the underlying array that hold the amplitudes): 
```
array: [ a,  b,  c,  d ]
         0   1   2   3  -> indices as decimals
        00  01  10  11  -> indices as binary
        |0> |2> |1> |3> -> state as interpreted by the users
```
So, when the users does `s[1]`,  they are asking for `s["|10>"]`, and thus we need to translate `|1>` to the corresponding index into the vector of amplitudes. The high-level understanding of this translations is: we take `1`, translate it to `10`, and interpret that as a binary number, i.e., `2`. Then `s[1] == c`.

Reply:
The original intent was that the method would act like `s[1] = b`, give the user a view on the vector data. Granted, you are right that this may not apply to other state representations like those from cuTensorNet. If we move to this second interpretation, I would think we should put forward a better method to extract amplitudes (e.g. see discussion on #1107)

Reply:
I would prefer we keep it as is - there are methods for direct vector and matrix extraction. We should provide the correct updates as part of #1123 to extract amplitudes more generally as discussed in #1107

Reply:
Talked offline. We are picking an amplitude ordering (following cuQuantum's) which is a good thing - enforces backend simulators to be compliant, helps future compiler passes. But here we are retaining the current amplitude ordering for the user-facing get_state function. It would be preferable for the get_state function to return a state of the same bit ordering as the internal system (cuQuantum's after this PR) but this is a "breaking change". To get this PR across the finish line, we'll retain this status quo, but future work should allow users to specify the bit ordering when calling get_state, and ultimately the default should be the same as the simulation infrastructure.